File: blind-75/word_search_II.py
# leetcode 212

from typing import List
from collections import defaultdict

class Solution:
	def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:

		def backtrackAndBuilWords(board, wordSet, visitSet, pos, curr) -> None:

			i, j = pos
			curr.append(board[i][j])
			wordSet.add("".join(curr))

			tempChar = board[i][j]
			board[i][j] = "#"

			for x, y in [[0, -1], [-1, 0], [0, 1], [1, 0]]:
				dx, dy = i + x, j + y
				if 0 <= dx < len(board) and 0 <= dy < len(board[0]) and \
						board[dx][dy] != "#" and (dx, dy) not in visitSet:
					# visitSet.add((dx, dy))
					backtrackAndBuilWords(board, wordSet, visitSet, (dx, dy), curr)
					# visitSet.remove((dx, dy))

			board[i][j] = tempChar


		wordSet = set()
		visitSet = set()

		for i in range(len(board)):
			for j in range(len(board[0])):
				visitSet.clear()
				backtrackAndBuilWords(board, wordSet, visitSet, (i, j), [])

		print(wordSet)
		res = []
		for word in words:
			if word in wordSet:
				res.append(word)

		return res


obj = Solution()
board = [["a","b"],["c","d"]]
words = ["abcb"]
print(obj.findWords(board, words))
# board = [["o","a","a","n"],["e","t","a","e"],["i","h","k","r"],["i","f","l","v"]]
# words = ["oath","pea","eat","rain"]
# print(obj.findWords(board, words))

File: blind-75/word_search_II.java
// leetcode 212

import java.util.*;

class Solution {
	public void dfsTraverse(int i, int j, int rowSize, int colSize, char[][] board, Stack<String> pathString, Set<String> wordSet) {
		pathString.add(String.valueOf(board[i][j]));

		String word = "";
		for (String s: pathString) {
			word += s;
		}
		wordSet.add(word);
		// System.out.println("current pathString " + pathString);
		// System.out.println("current wordSet " + wordSet);

		char tempBoardValue = board[i][j];
		board[i][j] = '#';

		int[][] moves = {{0, -1}, {-1, 0}, {0, 1}, {1, 0}};
		for (int[] move : moves) {
			int dx = i + move[0];
			int dy = j + move[1];

			if ((dx >= 0 && dx < rowSize) && (dy >= 0 && dy < colSize) && (board[dx][dy] != '#')) {
				dfsTraverse(dx, dy, rowSize, colSize, board, pathString, wordSet);
			}
		}

		pathString.pop();
		board[i][j] = tempBoardValue;
	}

	public List<String> findWords(char[][] board, String[] words) {
		int rowSize = board.length;
		int colSize = board[0].length;
		List<String> res = new ArrayList<>();

		Set<String> wordSet = new HashSet<>();
		Stack<String> pathString = new Stack<>();

		for (int i = 0; i < rowSize; i++) {
			for (int j = 0; j < colSize; j++) {
				pathString.clear();
				dfsTraverse(i, j, rowSize, colSize, board, pathString, wordSet);
			}
		}

		// System.out.println("final wordSet " + wordSet);
		for (String word: words) {
			if (wordSet.contains(word)) {
				res.add(word);
			}
		}

		Collections.sort(res);
		return res;
	}
}


class Main {
	public static void main(String[] args) {
		Solution obj = new Solution();

		char[][] board = {{'o','a','a','n'},{'e','t','a','e'},{'i','h','k','r'},{'i','f','l','v'}};
		String[] words = {"oath","pea","eat","rain"};

		// char[][] board = {{'a', 'b'}, {'c', 'd'}};
		// String[] words = {"abcd"};
		System.out.println(obj.findWords(board, words));
	}
}